### PR TITLE
Gen4: Improve merging of DBA queries

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description
This PR adds support to merge DBA queries in Gen4 if they are being routed to the same keyspace

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/querytree_transformers.go [hunk 1]
```diff
@@ -309,8 +315,15 @@ func mergeUnionLogicalPlans(ctx *planningContext, left logicalPlan, right logica
...
 
+	// safe to append system table schema and system table names, since either the routing will match or either side would be throwing an error
+	// during run-time which we want to preserve. For example outer side has User in sys table schema and inner side has User and Main in sys table schema
+	// Inner might end up throwing an error at runtime, but if it doesn't then it is safe to merge.
 	if canMergeUnionPlans(ctx, lroute, rroute) {
 		lroute.Select = &sqlparser.Union{Left: lroute.Select, Distinct: false, Right: rroute.Select}
+		lroute.eroute.SysTableTableSchema = append(lroute.eroute.SysTableTableSchema, rroute.eroute.SysTableTableSchema...)
+		for k, v := range rroute.eroute.SysTableTableName {
+			lroute.eroute.SysTableTableName[k] = v
+		}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/querytree_transformers.go:326]: nit: we can export this and lines 149 to 152 to a method

**Review by frouioui — APPROVED** [review]
LGTM

File: go/vt/vtgate/planbuilder/querytree_transformers.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/querytree_transformers.go]: Fixed

File: go/vt/vtgate/planbuilder/horizon_planning.go [hunk 2]
```diff
@@ -316,6 +316,18 @@ func pushProjection(expr *sqlparser.AliasedExpr, plan logicalPlan, semTable *sem
...
 		return len(node.cols) - 1, true, nil
+	case *concatenateGen4:
+		offset, added, err := pushProjection(expr, node.sources[0], semTable, inner, reuseCol, hasAggregation)
+		if err != nil {
+			return 0, false, err
+		}
+		if added {
+			if hasAggregation {
+				return 0, false, vterrors.New(vtrpcpb.Code_UNIMPLEMENTED, "aggregation not supported on top of unions")
+			}
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/horizon_planning.go:327]: we can fail this on top itself. WDYT?

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 3]
```diff
@@ -1019,6 +1012,39 @@ func canMergeSubqueryPlans(ctx *planningContext, a, b *route) bool {
...
+		}
+	}
+
+	// when we have no information from either side, they both will be routed to default keyspace, so they can be merged
+	return len(a.eroute.SysTableTableSchema) == 0 &&
+		len(a.eroute.SysTableTableName) == 0 &&
+		len(b.eroute.SysTableTableSchema) == 0 &&
+		len(b.eroute.SysTableTableName) == 0
+}
+
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/route_planning.go:1047]: we should also be able to merge if either side is empty. i.e. does not contain sytable table name and schema.

File: go/vt/vtgate/planbuilder/horizon_planning.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/horizon_planning.go:327]: also, add a test for it.

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**